File: apps/android/app/src/main/java/com/cursumi/app/ui/courses/QuizViews.kt
package com.cursumi.app.ui.courses

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateMapOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.navigation.NavController
import com.cursumi.app.core.model.Exam
import com.cursumi.app.core.model.ExamResult
import com.cursumi.app.core.model.Lesson
import com.cursumi.app.core.model.QuizAnswer
import com.cursumi.app.core.model.QuizConfig
import com.cursumi.app.core.model.QuizQuestion
import com.cursumi.app.core.model.QuizQuestionType
import com.cursumi.app.core.ui.Brand
import com.cursumi.app.core.ui.BrandTextField
import com.cursumi.app.core.ui.EmptyState
import com.cursumi.app.core.ui.ErrorText
import com.cursumi.app.core.ui.GhostButton
import com.cursumi.app.core.ui.Loading
import com.cursumi.app.core.ui.PrimaryButton
import com.cursumi.app.core.ui.Screen
import com.cursumi.app.ui.LocalApp
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

private fun allAnswered(questions: List<QuizQuestion>, answers: Map<Int, QuizAnswer>) =
    questions.indices.all { questions[it].type == QuizQuestionType.ORDERING || (answers[it]?.isAnswered == true) }

/** Quiz de lección: se califica en cliente y se guarda con `completeLesson`. */
@Composable
fun LessonQuizView(lesson: Lesson, onCompleted: (String) -> Unit) {
    val app = LocalApp.current
    val scope = rememberCoroutineScope()
    val questions = remember(lesson.id) { QuizQuestion.parseLessonQuiz(lesson.content) }
    val config = remember(lesson.id) { QuizConfig.parse(lesson.content) }
    val answers = remember { mutableStateMapOf<Int, QuizAnswer>() }
    var submitted by remember { mutableStateOf(false) }
    var score by remember { mutableIntStateOf(0) }
    var saving by remember { mutableStateOf(false) }
    var attempts by remember { mutableIntStateOf(0) }
    var timeLeft by remember { mutableStateOf(if (config.timeLimitMin > 0) config.timeLimitMin * 60 else null) }

    val passed = score >= config.passingScore
    val attemptsLeft = if (config.maxAttempts > 0) maxOf(0, config.maxAttempts - attempts) else null
    val canRetake = !passed && (config.maxAttempts == 0 || attempts < config.maxAttempts)

    fun submit() {
        if (submitted) return
        val correct = questions.indices.count { questions[it].grade(answers[it]) }
        score = if (questions.isEmpty()) 0 else Math.round(correct * 100.0 / questions.size).toInt()
        submitted = true; timeLeft = null; attempts++
        scope.launch {
            saving = true
            runCatching { app.student.completeLesson(lesson.id, lesson.courseId, score, answers.mapKeys { it.key.toString() }) }
                .onSuccess { onCompleted(lesson.id) }
            saving = false
        }
    }

    // Cuenta regresiva: al llegar a 0 envía automáticamente.
    LaunchedEffect(timeLeft, submitted) {
        val t = timeLeft ?: return@LaunchedEffect
        if (submitted) return@LaunchedEffect
        if (t <= 0) { submit(); return@LaunchedEffect }
        delay(1000); timeLeft = t - 1
    }

    if (questions.isEmpty()) { Text("Este quiz no tiene preguntas.", fontStyle = FontStyle.Italic, color = Brand.muted); return }
    Column(verticalArrangement = Arrangement.spacedBy(14.dp)) {
        timeLeft?.let { if (!submitted) Column(Modifier.fillMaxWidth(), horizontalAlignment = Alignment.CenterHorizontally) { TimerPill(it) } }
        if (submitted) ResultCard(
            score, passed, if (passed) "¡Aprobado!" else "No alcanzaste el ${config.passingScore}% mínimo.",
            attemptsLeft?.takeIf { !passed }?.let { if (it > 0) "Intentos restantes: $it" else "Sin intentos restantes." },
        )
        questions.forEachIndexed { i, q ->
            QuestionCard(i + 1, q.question, if (submitted) q.grade(answers[i]) else null) {
                QuizAnswerInput(q.type, q.options, q.matchRight ?: emptyList(), answers[i], { answers[i] = it }, disabled = submitted)
            }
        }
        when {
            !submitted -> PrimaryButton("Enviar respuestas", enabled = allAnswered(questions, answers)) { submit() }
            saving -> Loading()
            canRetake -> GhostButton("Reintentar") { answers.clear(); submitted = false; score = 0; timeLeft = if (config.timeLimitMin > 0) config.timeLimitMin * 60 else null }
            else -> Text("Respuestas guardadas ✓", color = Brand.muted, modifier = Modifier.fillMaxWidth(), textAlign = androidx.compose.ui.text.style.TextAlign.Center)
        }
    }
}

/** Quiz de sección: lo califica el servidor. */
@Composable
fun SectionQuizView(lesson: Lesson, onCompleted: (String) -> Unit) {
    val app = LocalApp.current
    val scope = rememberCoroutineScope()
    val questions = remember(lesson.id) { QuizQuestion.parseSectionQuiz(lesson.sectionQuiz) }
    val answers = remember { mutableStateMapOf<Int, QuizAnswer>() }
    var submitted by remember { mutableStateOf(false) }
    var result by remember { mutableStateOf<Pair<Int, Boolean>?>(null) }
    var saving by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }

    if (questions.isEmpty()) { Text("Esta actividad no tiene preguntas.", fontStyle = FontStyle.Italic, color = Brand.muted); return }
    Column(verticalArrangement = Arrangement.spacedBy(14.dp)) {
        result?.let { ResultCard(it.first, it.second, if (it.second) "¡Aprobado!" else "Inténtalo de nuevo.") }
        questions.forEachIndexed { i, q ->
            QuestionCard(i + 1, q.question, if (submitted) q.grade(answers[i]) else null) {
                QuizAnswerInput(q.type, q.options, q.matchRight ?: emptyList(), answers[i], { answers[i] = it }, disabled = submitted)
            }
        }
        ErrorText(error)
        when {
            !submitted -> PrimaryButton("Enviar", enabled = allAnswered(questions, answers)) {
                val sectionId = lesson.sectionId
                if (sectionId == null) { error = "No se pudo identificar la sección."; return@PrimaryButton }
                submitted = true; saving = true; error = null
                scope.launch {
                    try { result = app.student.submitSectionQuiz(sectionId, lesson.courseId, answers.mapKeys { it.key.toString() }); onCompleted(lesson.id) }
                    catch (e: Exception) { error = e.message; submitted = false }
                    saving = false
                }
            }
            saving -> Loading()
            result?.second == false -> GhostButton("Reintentar") { answers.clear(); submitted = false; result = null }
        }
    }
}

/** Tarea: respuesta de texto que se puede actualizar. */
@Composable
fun AssignmentView(lesson: Lesson, onCompleted: (String) -> Unit) {
    val app = LocalApp.current
    val scope = rememberCoroutineScope()
    var text by remember { mutableStateOf("") }
    var loading by remember { mutableStateOf(true) }
    var saving by remember { mutableStateOf(false) }
    var savedAt by remember { mutableStateOf<String?>(null) }
    var error by remember { mutableStateOf<String?>(null) }

    LaunchedEffect(lesson.id) {
        runCatching { app.student.assignment(lesson.id, lesson.courseId) }.getOrNull()?.let { text = it.content; savedAt = it.submittedAt }
        loading = false
    }
    if (loading) { Loading(); return }
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Text("Tu respuesta", fontWeight = FontWeight.SemiBold)
        BrandTextField(text, { text = it }, "Escribe tu respuesta", singleLine = false, minLines = 5)
        savedAt?.let { Text("Enviada · ${it.take(10)}", color = Brand.success, style = MaterialTheme.typography.labelSmall) }
        ErrorText(error)
        PrimaryButton(if (savedAt == null) "Enviar tarea" else "Actualizar respuesta", loading = saving) {
            val content = text.trim()
            if (content.isEmpty()) { error = "Escribe tu respuesta antes de enviar."; return@PrimaryButton }
            scope.launch {
                saving = true; error = null
                try {
                    app.student.submitAssignment(lesson.id, lesson.courseId, content)
                    app.student.completeLesson(lesson.id, lesson.courseId)
                    savedAt = java.time.Instant.now().toString(); onCompleted(lesson.id)
                } catch (e: Exception) { error = "No se pudo enviar. Inténtalo de nuevo." }
                saving = false
            }
        }
    }
}

/** Examen final del curso; el servidor califica y emite el certificado. */
@Composable
fun ExamScreen(nav: NavController, courseId: String) {
    val app = LocalApp.current
    val scope = rememberCoroutineScope()
    var exam by remember { mutableStateOf<Exam?>(null) }
    var loading by remember { mutableStateOf(true) }
    var error by remember { mutableStateOf<String?>(null) }
    val answers = remember { mutableStateMapOf<String, QuizAnswer>() }
    var result by remember { mutableStateOf<ExamResult?>(null) }
    var submitting by remember { mutableStateOf(false) }
    var timeLeft by remember { mutableStateOf<Int?>(null) }

    LaunchedEffect(courseId) {
        try { exam = app.student.exam(courseId); exam?.timeLimit?.takeIf { it > 0 }?.let { timeLeft = it * 60 } }
        catch (e: Exception) { error = "No se pudo cargar el examen." }
        loading = false
    }

    fun submit() {
        if (exam == null || result != null || submitting) return
        scope.launch {
            submitting = true; error = null
            try { result = app.student.submitExam(courseId, answers.toMap()); timeLeft = null } catch (e: Exception) { error = e.message }
            submitting = false
        }
    }

    LaunchedEffect(timeLeft, result) {
        val t = timeLeft ?: return@LaunchedEffect
        if (result != null) return@LaunchedEffect
        if (t <= 0) { submit(); return@LaunchedEffect }
        delay(1000); timeLeft = t - 1
    }

    Screen("Examen", onBack = { nav.popBackStack() }) { padding ->
        val e = exam
        when {
            loading -> Loading()
            e == null -> EmptyState(error ?: "Este curso no tiene examen final.")
            else -> Column(Modifier.fillMaxSize().padding(padding).verticalScroll(rememberScrollState()).padding(16.dp), verticalArrangement = Arrangement.spacedBy(14.dp)) {
                Text("Examen final", style = MaterialTheme.typography.headlineSmall, fontWeight = FontWeight.Bold)
                Text("Necesitas ${e.passingScore}% para aprobar.", color = Brand.muted)
                timeLeft?.let { if (result == null) Column(Modifier.fillMaxWidth(), horizontalAlignment = Alignment.CenterHorizontally) { TimerPill(it) } }
                result?.let { r -> ResultCard(Math.round(r.score).toInt(), r.passed, extra = r.certificate?.let { "🎓 Certificado emitido · Folio ${it.number}" }) }
                e.questions.forEachIndexed { i, q ->
                    QuestionCard(i + 1, q.question) {
                        if (q.isGradable) QuizAnswerInput(q.questionType, q.options ?: emptyList(), q.matchRight ?: emptyList(), answers[q.id], { answers[q.id] = it }, disabled = result != null)
                        else Text("Pregunta abierta — se evalúa manualmente.", fontStyle = FontStyle.Italic, color = Brand.muted, style = MaterialTheme.typography.labelSmall)
                    }
                }
                ErrorText(error)
                if (result == null) {
                    val gradable = e.questions.filter { it.isGradable }
                    val ok = gradable.all { it.questionType == QuizQuestionType.ORDERING || answers[it.id]?.isAnswered == true }
                    PrimaryButton("Enviar examen", loading = submitting, enabled = ok) { submit() }
                }
            }
        }
    }
}

File: apps/ios/Cursumi/Features/MyCourses/QuizViews.swift
import SwiftUI

/// Quiz de lección: se califica en cliente y se guarda con `completeLesson`.
struct LessonQuizView: View {
    let lesson: Lesson
    let onCompleted: (String) -> Void

    @State private var questions: [QuizQuestion] = []
    @State private var config = QuizConfig()
    @State private var answers: [Int: QuizAnswer] = [:]
    @State private var submitted = false
    @State private var score = 0
    @State private var saving = false
    @State private var attempts = 0
    @State private var timeLeft: Int?
    private let api = StudentAPI()

    private var passed: Bool { score >= config.passingScore }
    private var attemptsLeft: Int? { config.maxAttempts > 0 ? max(0, config.maxAttempts - attempts) : nil }
    private var canRetake: Bool { !passed && (config.maxAttempts == 0 || attempts < config.maxAttempts) }
    private var allAnswered: Bool {
        questions.indices.allSatisfy { questions[$0].type == .ordering || (answers[$0]?.isAnswered ?? false) }
    }

    var body: some View {
        VStack(spacing: 14) {
            if questions.isEmpty {
                Text("Este quiz no tiene preguntas.").italic().foregroundStyle(.secondary)
            } else {
                if let timeLeft, !submitted {
                    TimerPill(seconds: timeLeft)
                }
                if submitted {
                    ResultCard(
                        score: score, passed: passed,
                        message: passed ? "¡Aprobado!" : "No alcanzaste el \(config.passingScore)% mínimo.",
                        extra: attemptsLeft.flatMap { passed ? nil : ($0 > 0 ? "Intentos restantes: \($0)" : "Sin intentos restantes.") }
                    )
                }
                ForEach(questions.indices, id: \.self) { i in
                    QuestionCard(number: i + 1, text: questions[i].question, graded: submitted ? questions[i].grade(answers[i]) : nil) {
                        QuizAnswerInput(
                            type: questions[i].type, options: questions[i].options,
                            matchRight: questions[i].matchRight ?? [],
                            answer: binding(i), disabled: submitted
                        )
                    }
                }
                if !submitted {
                    PrimaryButton(title: "Enviar respuestas", disabled: !allAnswered) { Task { await submit() } }
                } else if saving {
                    ProgressView()
                } else if canRetake {
                    GhostButton(title: "Reintentar") { retake() }
                } else {
                    Text("Respuestas guardadas ✓").foregroundStyle(.secondary)
                }
            }
        }
        .onAppear {
            questions = QuizQuestion.parseLessonQuiz(lesson.content)
            config = QuizConfig.parse(lesson.content)
            timeLeft = config.timeLimitMin > 0 ? config.timeLimitMin * 60 : nil
        }
        .task(id: timeLeft) {
            // Cuenta regresiva: al llegar a 0 envía automáticamente.
            guard let t = timeLeft, !submitted else { return }
            if t <= 0 { await submit(); return }
            try? await Task.sleep(for: .seconds(1))
            if !Task.isCancelled, timeLeft == t { timeLeft = t - 1 }
        }
    }

    private func binding(_ i: Int) -> Binding<QuizAnswer?> {
        Binding(get: { answers[i] }, set: { answers[i] = $0 })
    }

    private func submit() async {
        guard !submitted else { return }
        let correct = questions.indices.filter { questions[$0].grade(answers[$0]) }.count
        score = questions.isEmpty ? 0 : Int((Double(correct) / Double(questions.count) * 100).rounded())
        submitted = true
        timeLeft = nil
        attempts += 1
        saving = true
        defer { saving = false }
        var toSave: [String: QuizAnswer] = [:]
        for (i, a) in answers { toSave[String(i)] = a }
        do {
            try await api.completeLesson(lesson.id, courseId: lesson.courseId, score: score, answers: toSave)
            onCompleted(lesson.id)
        } catch {
            // El puntaje ya se muestra; el intento siguiente vuelve a guardar.
        }
    }

    private func retake() {
        answers = [:]
        submitted = false
        score = 0
        timeLeft = config.timeLimitMin > 0 ? config.timeLimitMin * 60 : nil
    }
}

/// Quiz de sección: lo califica el servidor.
struct SectionQuizView: View {
    let lesson: Lesson
    let onCompleted: (String) -> Void

    @State private var questions: [QuizQuestion] = []
    @State private var answers: [Int: QuizAnswer] = [:]
    @State private var submitted = false
    @State private var result: (score: Int, passed: Bool)?
    @State private var saving = false
    @State private var error: String?
    private let api = StudentAPI()

    private var allAnswered: Bool {
        questions.indices.allSatisfy { questions[$0].type == .ordering || (answers[$0]?.isAnswered ?? false) }
    }

    var body: some View {
        VStack(spacing: 14) {
            if questions.isEmpty {
                Text("Esta actividad no tiene preguntas.").italic().foregroundStyle(.secondary)
            } else {
                if let result {
                    ResultCard(score: result.score, passed: result.passed, message: result.passed ? "¡Aprobado!" : "Inténtalo de nuevo.")
                }
                ForEach(questions.indices, id: \.self) { i in
                    QuestionCard(number: i + 1, text: questions[i].question, graded: submitted ? questions[i].grade(answers[i]) : nil) {
                        QuizAnswerInput(
                            type: questions[i].type, options: questions[i].options,
                            matchRight: questions[i].matchRight ?? [],
                            answer: Binding(get: { answers[i] }, set: { answers[i] = $0 }), disabled: submitted
                        )
                    }
                }
                if let error { Text(error).foregroundStyle(Brand.danger) }
                if !submitted {
                    PrimaryButton(title: "Enviar", disabled: !allAnswered) { Task { await submit() } }
                } else if saving {
                    ProgressView()
                } else if let result, !result.passed {
                    GhostButton(title: "Reintentar") { answers = [:]; submitted = false; self.result = nil }
                }
            }
        }
        .onAppear { questions = QuizQuestion.parseSectionQuiz(lesson.sectionQuiz) }
    }

    private func submit() async {
        guard let sectionId = lesson.sectionId else { error = "No se pudo identificar la sección."; return }
        submitted = true
        saving = true
        error = nil
        defer { saving = false }
        var payload: [String: QuizAnswer] = [:]
        for (i, a) in answers { payload[String(i)] = a }
        do {
            result = try await api.submitSectionQuiz(sectionId: sectionId, courseId: lesson.courseId, answers: payload)
            onCompleted(lesson.id)
        } catch {
            self.error = error.localizedDescription
            submitted = false
        }
    }
}

/// Tarea: respuesta de texto que se puede actualizar.
struct AssignmentView: View {
    let lesson: Lesson
    let onCompleted: (String) -> Void

    @State private var text = ""
    @State private var loading = true
    @State private var saving = false
    @State private var savedAt: String?
    @State private var error: String?
    private let api = StudentAPI()

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            if loading {
                ProgressView().frame(maxWidth: .infinity)
            } else {
                Text("Tu respuesta").font(.subheadline.weight(.semibold))
                TextEditor(text: $text)
                    .frame(minHeight: 140)
                    .padding(8)
                    .overlay(RoundedRectangle(cornerRadius: 12).stroke(Color.gray.opacity(0.3)))
                if let savedAt {
                    Text("Enviada · \(savedAt.prefix(10))").font(.caption).foregroundStyle(Brand.success)
                }
                if let error { Text(error).foregroundStyle(Brand.danger) }
                PrimaryButton(title: savedAt == nil ? "Enviar tarea" : "Actualizar respuesta", loading: saving) {
                    Task { await submit() }
                }
            }
        }
        .task {
            if let sub = try? await api.assignment(lessonId: lesson.id, courseId: lesson.courseId) {
                text = sub.content
                savedAt = sub.submittedAt
            }
            loading = false
        }
    }

    private func submit() async {
        let content = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !content.isEmpty else { error = "Escribe tu respuesta antes de enviar."; return }
        error = nil
        saving = true
        defer { saving = false }
        do {
            try await api.submitAssignment(lessonId: lesson.id, courseId: lesson.courseId, content: content)
            try await api.completeLesson(lesson.id, courseId: lesson.courseId)
            savedAt = ISO8601DateFormatter().string(from: Date())
            onCompleted(lesson.id)
        } catch {
            self.error = "No se pudo enviar. Inténtalo de nuevo."
        }
    }
}

/// Examen final del curso; el servidor califica y emite el certificado.
struct ExamView: View {
    let courseId: String

    @State private var exam: Exam?
    @State private var loading = true
    @State private var error: String?
    @State private var answers: [String: QuizAnswer] = [:]
    @State private var result: ExamResult?
    @State private var submitting = false
    @State private var timeLeft: Int?
    private let api = StudentAPI()

    private var allAnswered: Bool {
        (exam?.questions ?? []).filter(\.isGradable).allSatisfy { $0.type == .ordering || (answers[$0.id]?.isAnswered ?? false) }
    }

    var body: some View {
        Group {
            if loading {
                ProgressView()
            } else if let exam {
                ScrollView {
                    VStack(alignment: .leading, spacing: 14) {
                        Text("Examen final").font(.title2.bold())
                        Text("Necesitas \(exam.passingScore)% para aprobar.").foregroundStyle(.secondary)
                        if let timeLeft, result == nil { TimerPill(seconds: timeLeft) }
                        if let result {
                            ResultCard(
                                score: Int(result.score.rounded()), passed: result.passed,
                                extra: result.certificate.map { "🎓 Certificado emitido · Folio \($0.number)" }
                            )
                        }
                        ForEach(Array(exam.questions.enumerated()), id: \.element.id) { i, q in
                            QuestionCard(number: i + 1, text: q.question) {
                                if q.isGradable {
                                    QuizAnswerInput(
                                        type: q.type ?? .multipleChoice, options: q.options ?? [], matchRight: q.matchRight ?? [],
                                        answer: Binding(get: { answers[q.id] }, set: { answers[q.id] = $0 }),
                                        disabled: result != nil
                                    )
                                } else {
                                    Text("Pregunta abierta — se evalúa manualmente.").font(.caption).italic().foregroundStyle(.secondary)
                                }
                            }
                        }
                        if let error { Text(error).foregroundStyle(Brand.danger) }
                        if result == nil {
                            PrimaryButton(title: "Enviar examen", loading: submitting, disabled: !allAnswered) { Task { await submit() } }
                        }
                    }
                    .padding(16)
                }
            } else {
                EmptyState(title: error ?? "Este curso no tiene examen final.")
            }
        }
        .navigationTitle("Examen")
        .navigationBarTitleDisplayMode(.inline)
        .task {
            do {
                exam = try await api.exam(courseId: courseId)
                if let limit = exam?.timeLimit, limit > 0 { timeLeft = limit * 60 }
            } catch {
                self.error = "No se pudo cargar el examen."
            }
            loading = false
        }
        .task(id: timeLeft) {
            guard let t = timeLeft, result == nil else { return }
            if t <= 0 { await submit(); return }
            try? await Task.sleep(for: .seconds(1))
            if !Task.isCancelled, timeLeft == t { timeLeft = t - 1 }
        }
    }

    private func submit() async {
        guard exam != nil, result == nil, !submitting else { return }
        submitting = true
        error = nil
        defer { submitting = false }
        do {
            result = try await api.submitExam(courseId: courseId, answers: answers)
            timeLeft = nil
        } catch {
            self.error = error.localizedDescription
        }
    }
}

struct TimerPill: View {
    let seconds: Int
    var body: some View {
        Label(seconds.mmss, systemImage: "timer")
            .font(.subheadline.bold())
            .foregroundStyle(seconds <= 30 ? Brand.danger : .primary)
            .padding(.horizontal, 16).padding(.vertical, 6)
            .overlay(Capsule().stroke(seconds <= 30 ? Brand.danger : Color.gray.opacity(0.3)))
            .frame(maxWidth: .infinity)
    }
}
